Feat #25150: Pose Graph MST initialization

03kiko opened the pull request:
Implements an MST-based initialisation for pose graphs, as proposed in issue #25150. Both Prim’s and Kruskal’s algorithms were added to the 3D module. These receive a vector of node IDs and a vector of edges (each with source and target IDs and a weight), and return a vector with the resulting edges. These MST implementations treat edges as undirected internally, meaning users only need to provide one direction (A→B or B→A), and duplicates are handled automatically.

Additionally, a new pose graph initialisation method using MST (Prim) was implemented. It constructs the MST over the pose graph, then traverses it to reconstruct node poses. With this, users can call `poseGraph->initializePosesWithMST()` to create an initial solution for the pose graph problem.

A set of test cases validating the implementation was also included.

#### Notes
- The edge weight used in the MST for pose graphs is calculated as:
`weight = || translation || + λ * rotation_angle`,
where λ = 0.485 was determined empirically based on optimiser performance;
- Validated on [Sphere-a](https://lucacarlone.mit.edu/datasets/) pose graph, showing similar convergence behaviour with or without MST initialisation;
- Alternative weight formulas, such as the Mahalanobis distance formula, were also tested, but the current formula yielded better results.

#### Future Work
- Extend testing to more diverse pose graphs (currently limited to [Sphere-a](https://github.com/opencv/opencv_extra/blob/5.x/testdata/cv/rgbd/sphere_bignoise_vertex3.g2o) in opencv_extra)
- Explore adaptive tuning of the λ parameter for broader applicability.

Co-authored-by: @miguel1099 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented on modules/3d/include/opencv2/3d/detail/mst.hpp:
```diff
@@ -0,0 +1,38 @@
+};
+
+// builds a MST using Prim's algorithm
+CV_EXPORTS std::vector<MSTEdge> buildMSTPrim(
+    const size_t numNodes,
+    const std::vector<MSTEdge>& edges,
+    size_t root = 0
+);
+
+// builds a MST using Kruskal's algorithm
+CV_EXPORTS std::vector<MSTEdge> buildMSTKruskal(
+    const size_t numNodes,
+    const std::vector<MSTEdge>& edges
```
I propose to use single function as public API with algorithm type parameter like it's done in many places in OpenCV. Also we have internal MST implementation at least in multi-camera calibration. It makes sense to move it to main header, not details.

03kiko commented:
Hi @asmorkalov, thanks for the feedback! Please let us know if everything looks good now or if there’s anything else you'd like us to adjust!

03kiko commented:
Hi @asmorkalov, could you trigger CI again? It seems to have failed to clone opencv_extra?

asmorkalov commented on modules/3d/include/opencv2/3d/mst.hpp:
```diff
@@ -0,0 +1,36 @@
+};
+
+// builds a MST using the selected algorithm (Prim or Kruskal).
+CV_EXPORTS std::vector<MSTEdge> buildMST(
```
CV_EXPORTS_W

asmorkalov replied:
The graph may have 2 or more isolated pieces. I propose the following:
1. return status, e.g. bool
2. add `CV_OUT std::vector<MSTEdge>&` parameter for result.

Also the solution will be more friendly solution for Python and Java bindings.

asmorkalov commented on modules/3d/include/opencv2/3d/mst.hpp:
```diff
@@ -0,0 +1,36 @@
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#ifndef OPENCV_3D_DETAIL_MST_HPP
+#define OPENCV_3D_DETAIL_MST_HPP
```
remove _DETAIL

asmorkalov commented on modules/3d/include/opencv2/3d/mst.hpp:
```diff
@@ -0,0 +1,36 @@
+namespace cv
+{
+
+// represents an edge in a graph for MST computation
```
Please use doxygen notation for the documentation.

03kiko commented:
Hi @asmorkalov, @opencv-alalek, we've just pushed a new version addressing your reviews. Please let me know if everything looks good or if there's anything you'd like us to adjust.

asmorkalov commented on modules/3d/src/mst.cpp:
```diff
@@ -0,0 +1,160 @@
+        case MST_KRUSKAL:
+            result = buildMSTKruskal(numNodes, inputEdges, resultingEdges);
+            break;
+        default:
+            CV_LOG_INFO(NULL, "Invalid MST algorithm specified");
+            result = false;
```
It's definitely invalid user input. It should not be just ignored. We use CV_Assert / CV_Error and throw exception with it.

asmorkalov commented on modules/3d/src/mst.cpp:
```diff
@@ -0,0 +1,160 @@
+            result = false;
+    }
+
+    return (result && resultingEdges.size() >= static_cast<size_t>(numNodes - 1));
```
It's the only place where false is returned, besides the obvious one. What happens if the MST cannot be built, e.g. the graph consists of 2 split sub-graphs? I would say, it should return false too.

03kiko commented:
Hi @asmorkalov, just pushed a new version addressing your comments! Let me know if this looks good to you.

asmorkalov commented on modules/3d/src/rgbd/pose_graph.cpp:
```diff
@@ -472,6 +479,112 @@ bool PoseGraphImpl::isValid() const
+    double rotationAngle = cv::norm(rvec);
+
+    // empirically determined
+    double lambda = 0.485;
```
Looks like some data specific parameter. I propose to add lambda parameter with default value 0.485 to `initializePosesWithMST` to have opportunity to change it without OpenCV modification and rebuild. The value may be promoted to the weight calculator as parameter too.

asmorkalov commented on modules/3d/test/test_mst.cpp:
```diff
@@ -0,0 +1,267 @@
+    EXPECT_TRUE(resultKruskal);
+    EXPECT_EQ(primMST.size(), static_cast<size_t>(numNodes - 1));
+    EXPECT_EQ(kruskalMST.size(), static_cast<size_t>(numNodes - 1));
+}
```
Please add a couple of tests with negative scenarios:
- split graph of 2-3 sub-graphs
- one isolated node in the graph

asmorkalov commented on modules/3d/test/test_pose_graph.cpp:
```diff
@@ -155,6 +155,109 @@ TEST(PoseGraph, sphereG2O)
+    if (cvtest::debugLevel > 0)
+    {
+        // Write OBJ for MST-initialized pose graph with optimizer
+        {
+            std::string fname = "pg_with_mst_and_optimizer.obj";
+            std::fstream of(fname, std::fstream::out);
+            std::vector<size_t> ids = pgWithMSTAndOptimizer->getNodesIds();
+            for (const size_t& id : ids)
+            {
+                Point3d d = pgWithMSTAndOptimizer->getNodePose(id).translation();
+                of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+            }
+            size_t esz = pgWithMSTAndOptimizer->getNumEdges();
+            for (size_t i = 0; i < esz; i++)
+            {
+                size_t sid = pgWithMSTAndOptimizer->getEdgeStart(i), tid = pgWithMSTAndOptimizer->getEdgeEnd(i);
+                of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+            }
+            of.close();
+        }
```
We have saveMesh / savePointCloud for obj and ply i/o. Please use it: https://docs.opencv.org/5.x/da/d35/group____3d.html#gad565f2878b3d8c3f959ae7967a520bb2

asmorkalov commented on modules/3d/test/test_pose_graph.cpp:
```diff
@@ -155,6 +155,136 @@ TEST(PoseGraph, sphereG2O)
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = pgWithMSTAndOptimizer->getEdgeStart(i), tid = pgWithMSTAndOptimizer->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
+
+        // Write OBJ for optimizer-only pose graph
+        std::string fname = "pg_optimizer_only.obj";
+        std::fstream of(fname, std::fstream::out);
+        std::vector<size_t> ids = pgOptimizerOnly->getNodesIds();
+        for (const size_t& id : ids)
+        {
+            Point3d d = pgOptimizerOnly->getNodePose(id).translation();
+            of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+        }
+        size_t esz = pgOptimizerOnly->getNumEdges();
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = pgOptimizerOnly->getEdgeStart(i), tid = pgOptimizerOnly->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
+
+        // Write OBJ for initial pose graph
+        std::string fname = "pg_init.obj";
+        std::fstream of(fname, std::fstream::out);
+        std::vector<size_t> ids = init->getNodesIds();
+        for (const size_t& id : ids)
+        {
+            Point3d d = init->getNodePose(id).translation();
+            of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+        }
+        size_t esz = init->getNumEdges();
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = init->getEdgeStart(i), tid = init->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
```
Looks like this code is not needed any more, as you added implementation with `saveMesh` below.

asmorkalov commented:
Let's presume the original "hand made" version, but please add a comment why it differs from saveMesh. It's not obvious without detailed code analysis.
